### ISSUE TITLE
Support for "runtimeOnly" dependency configuration in C++

### ISSUE THREAD

**Issue by b-john** [gradle/gradle-native#831]
### Desired Behavior
C++ plugins support "runtimeOnly" dependency configuration where listed dependencies are only included as part of the runtime and are not compiled/linked against. 

### Current Behavior
Today, the java plugin supports runtimeOnly. 
- https://docs.gradle.org/current/userguide/java_library_plugin.html
- https://github.com/gradle/gradle/blob/7a3973819242c98612119694fefcb78ba4a94355/subprojects/plugins/src/main/java/org/gradle/api/plugins/JavaPlugin.java#L421

But the c++ plugins do not.
- https://github.com/gradle/gradle/blob/f868fb823df318d95f8b957f699afd03cd778d87/subprojects/language-native/src/main/java/org/gradle/language/cpp/internal/DefaultCppBinary.java#L75

### Context
We have libraries that are runtime-only dependencies. These dependencies are accessed at runtime with "dlopen" on Linux systems or "LoadLibrary" on Windows systems; We'd like to get those into the install directory without adding them to the linker line of the project.

This issue may be similar to https://github.com/gradle/gradle-native/issues/451, support for libraries without public headers. Thanks y'all for all the help/thoughts!

**Comment by lacasseio** [comment]
You are right to request this. C++ runtime dependencies are something that needs to be improved. However, it is possible, not convenient, to add a runtime dependency using the configuration with prefix `nativeRuntime`. We should start with a sample to demonstrate the current functionality and bridge the gap with this issue.

**Comment by b-john** [comment]
Agreed & thanks @lacasseio;  Apologies for the beginner question Daniel-- Do you have any guidance on how to append to the nativeRuntime configuration? 

I can add/create the configuration "nativeRuntime" but struggling to actually append to the existing one within DefaultCppBinary / looks like it does not get added to the main ConfigurationContainer.
Have read several articles including Sterling's: https://discuss.gradle.org/t/adding-dependencies-to-a-configuration-created-from-a-custom-source-set/8801/2 


This works as expected for creating a custom configuration but does not impact the nativeRuntime in DefaultCppBinary:
```groovy
configurations {
   nativeRuntime
}
dependencies {
   nativeRuntime project(':my_library')
}
```
Leaving out the declaration of "nativeRuntime" configuration results in the expected message around DependencyHandler not finding a function by that name / Sterling's response about it also failing to find the existing configuration on the fallback. 

Thanks again for your help & time!

**Comment by lacasseio** [comment]
The native samples repository has a similar example for what you are trying to achieve. Adding a dependency to the binary will be similar to [this code](https://github.com/gradle/native-samples/blob/master/cpp/prebuilt-binaries/build.gradle#L52-L64). In this sample, we add a prebuilt library, however, you can also add a dependency on another project.

**Comment by b-john** [comment]
Thanks @lacasseio ! Works like a charm; Able to add the following to our root module
```groovy
subprojects {
    ext.runtimeOnly = { Object dependencyNotation ->
        project.components.main.binaries.whenElementFinalized { binary ->
         binary.runtimeElements.get().dependencies.add(project.dependencies.create(dependencyNotation))
        }
    }
}
```
And then reference it across our modules as we would expect
```groovy
    dependencies {
        api project(':api project dependency')
        implementation project(':implementation project dependency')
        runtimeOnly project(':runtime-only project dependency')
    }
```
------------------------------
I am good with closing this issue given the functionality exists and isn't that hacky to leverage. Ideally there'd be more documentation / be exposed through the code instead of requiring that extra closure but that sounds like low priority. Thanks again Daniel, much appreciated.
